alacritty_terminal/src/tty/unix.rs

```diff
@@ -121,17 +118,38 @@ impl Pty {
     }
 }
 
-#[cfg(target_os = "macos")]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    let shell_name = pw.shell.rsplit('/').next().unwrap();
-    let argv = vec![String::from("-c"), format!("exec -a -{} {}", shell_name, pw.shell)];
+fn shell_command(shell_config: Option<&Program>, pw: &Passwd<'_>) -> Command {
+    // Look for a shell in the alacritty config first, then SHELL, then in passwd.
+    let shell = shell_config
+        .map(|prog| prog.program().to_owned())
+        .or_else(|| env::var("SHELL").ok())
+        .unwrap_or_else(|| pw.shell.to_owned());
 
-    Program::WithArgs { program: "/bin/bash".to_owned(), args: argv }
-}
+    #[cfg(not(target_os = "macos"))]
+    let mut command = Command::new(shell);
+
+    // On macOS, unless the user set the `shell` config in alacritty.yml, use the `login` command to
+    // launch the shell so programs like `w`, `who`, and `logname` work.
+    #[cfg(target_os = "macos")]
+    let mut command = if shell_config.is_some() {
+        Command::new(shell)
+    } else {
+        let mut login_command = Command::new("/usr/bin/login");
+        // -f bypasses authentication for the already-logged-in user
+        // -l makes it not a "login" shell and prevents login from changing directories to $HOME
+        // exec the shell with argv[0] prepended by - so the shell run as a login shell anyway
+        login_command.args(["-fl", pw.name, "/bin/zsh", "-c"]);
+        login_command.arg(format!("exec -a -{} {}", shell.rsplit('/').next().unwrap(), shell));
```

Comment:
Why is this split into a separate line, that just seems incredibly error-prone.

Also I'd like to keep the shell_name variable extracted, because that way it's going to be much more obvious what the whole split next unwrap crap is actually doing.

Reply:
It's split because `format` returns an owned string, while all the other arguments are borrowed strings, so I found it clearer and more concise to put them first rather than in one big array with `.to_owned` everywhere. I will refactor this a bit and keep the `shell_name` extracted.

Reply:
Extract the format, put it into the array as slice?

Reply:
Sure.